### ISSUE TITLE
Captured cursor mode

### ISSUE THREAD

**Issue by elmindreda** [glfw/glfw#58]
Add a cursor mode that restricts the cursor to the client area of the window.

**Comment by elmindreda** [comment]
This cannot be implemented on Wayland and is difficult to do on macOS. Closing this until the universe improves.

**Comment by linkmauve** [comment]
Actually this is already implemented on Wayland, since cb08dc574c55626460eb4860652f4de03fa2b0eb.

**Comment by elmindreda** [comment]
@linkmauve Wayland now allows constraining the system cursor to a window client area?

**Comment by linkmauve** [comment]
Yes, or even to a specific region of it. It’s currently used for GLFW_CURSOR_DISABLED in tandem with relative-pointer to prevent the pointer from leaving the client area.

**Comment by elmindreda** [comment]
Awesome.

**Comment by jadahl** [comment]
To clarify, what cb08dc5 does is lock the pointer using "wp_pointer_constraints.lock_pointer" so that it can no longer move while in 'disabled' mode. A 'captured' mode would instead use "wp_pointer_constraints.confine_pointer" which would confine it to a given region.

**Comment by IntellectualKitty** [comment]
As far as I know, on macOS you have to do this manually.

Inside the `- (void)mouseMoved:(NSEvent *)event` event handler, if the mouse position strays outside of the client area, you must first re-position it so that it is back inside the client area by calling `CGWarpMouseCursorPosition(updatedPosition)` and then call `CGAssociateMouseAndMouseCursorPosition(YES)` move the cursor to the updated mouse position.

Here's Apple's documentation for these functions:

https://developer.apple.com/reference/coregraphics/1456387-cgwarpmousecursorposition?language=objc
https://developer.apple.com/reference/coregraphics/1454486-cgassociatemouseandmousecursorpo?language=objc

**Comment by elmindreda** [comment]
It is reportedly possible using [event taps](https://developer.apple.com/reference/coregraphics/quartz_event_services), but only for non-store applications.

**Comment by SephiRok** [comment]
+1. Really important for strategy games / side-scrolling the camera.

**Comment by elmindreda** [comment]
@SephiRok It's my intent that this goes into 3.4, even if it means macOS gets left out initially.

**Comment by elmindreda** [comment]
- [ ] Windows
- [ ] X11
- [ ] macOS
- [ ] Wayland

**Comment by judos** [comment]
This has been fixed for some time, but would it be possible to release it in 3.3.9?
I guess 3.4 is still some time ahead.
